### PR TITLE
Fix the navigation property binding target path issue

### PR DESCRIPTION
**Pull request by xuzhg**
### Issues

*This pull request fixes issue #1938.*

### Description

*Briefly describe the changes of this pull request.*

### Checklist (Uncheck if it is not completed)

- [ ] *Test cases added*
- [ ] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

File: src/Microsoft.OData.Edm/ExtensionMethods/ExtensionMethods.cs [hunk 1]
```diff
@@ -3055,19 +3055,53 @@ internal static IEdmNavigationSource FindNavigationSourceExtended(this IEdmEntit
...
+            // the path could be:
+            // NS.Default.Customers/ContainedOrders  --> for back-compatible
+            // NS.Default/Customers/ContainedOrders
+            // Customers
+            // Customers/ContainedOrders
+            string[] pathSegments = path.Split('/');
+
+            string firstElementName = pathSegments[0];
+            int nextIndex = 1;
+            if (pathSegments[0].Contains("."))
```
Comment (gathogojr) [inline src/Microsoft.OData.Edm/ExtensionMethods/ExtensionMethods.cs:3070]:
```suggestion
            if (firstElementName.Contains("."))

```

File: src/Microsoft.OData.Edm/ExtensionMethods/ExtensionMethods.cs [hunk 2]
```diff
@@ -3055,19 +3055,53 @@ internal static IEdmNavigationSource FindNavigationSourceExtended(this IEdmEntit
...
+            {
+                if (pathSegments[0] == container.FullName())
```
Comment (gathogojr) [inline src/Microsoft.OData.Edm/ExtensionMethods/ExtensionMethods.cs:3072]:
```suggestion
                if (firstElementName == container.FullName())

```
```suggestion
                if (firstElementName == container.FullName())
```

Comment (gathogojr) [inline src/Microsoft.OData.Edm/ExtensionMethods/ExtensionMethods.cs:3072]: Might we want to do a case-insensitive comparison here?

**Review by gathogojr — APPROVED** [review]
Left a few comments otherwise LGTM

**Comment by gathogojr** [comment]
@xuzhg Based on your change, the generated metadata will have `NS/` in place of `NS.`. Your code takes care of parsing any Url in previous format and that seems sufficient to me but I'm not sure if Mike was implying that we should give the users control over the format of metadata to be generated in [this comment](https://github.com/OData/odata.net/issues/1938#issuecomment-740792461):
> For compatibility we should probably introduce a flag that controls whether the target is written in the old (incorrect) format or the new (correct) format. For 4.0 we should (unfortunately) make the old format for the default. For 4.01 metadata responses we can probably get away with changing the default to the new (correct) syntax.

File: src/Microsoft.OData.Edm/ExtensionMethods/ExtensionMethods.cs [hunk 3]
```diff
@@ -3055,19 +3055,53 @@ internal static IEdmNavigationSource FindNavigationSourceExtended(this IEdmEntit
         /// <returns>The navigation source found or empty if none found.</returns>
         internal static IEdmNavigationSource FindNavigationSource(this IEdmEntityContainer container, string path)
         {
-            string[] pathSegments = path.Split('.').Last().Split('/');
+            EdmUtil.CheckArgumentNull(container, "container");
+            EdmUtil.CheckArgumentNull(path, "path");
+
+            // the path could be:
+            // NS.Default.Customers/ContainedOrders  --> for back-compatible
```
Comment (mikepizzo) [inline src/Microsoft.OData.Edm/ExtensionMethods/ExtensionMethods.cs:3062]: or even NS.Default.Customers (for backward-compatibility) or NS.Default/Customers (for top-level entity set in the Default entity container)

File: src/Microsoft.OData.Edm/Schema/EdmEntitySet.cs [hunk 4]
```diff
@@ -41,7 +41,7 @@ public EdmEntitySet(IEdmEntityContainer container, string name, IEdmEntityType e
 
             this.container = container;
             this.type = new EdmCollectionType(new EdmEntityTypeReference(elementType, false));
-            this.path = new EdmPathExpression(this.container.FullName() + "." + this.Name);
+            this.path = new EdmPathExpression(this.container.FullName() + "/" + name);
```
Comment (mikepizzo) [inline src/Microsoft.OData.Edm/Schema/EdmEntitySet.cs:44]: >this.container.FullName() + "/" + name [](start = 46, length = 38)

use string.Concat(this.container.FullName(), "/", name)

File: src/Microsoft.OData.Edm/Schema/EdmSingleton.cs [hunk 5]
```diff
@@ -29,7 +29,7 @@ public EdmSingleton(IEdmEntityContainer container, string name, IEdmEntityType e
 
             this.container = container;
             this.entityType = entityType;
-            this.path = new EdmPathExpression(name);
+            this.path = new EdmPathExpression(this.container.FullName() + "/" + name);
```
Comment (mikepizzo) [inline src/Microsoft.OData.Edm/Schema/EdmSingleton.cs:32]: >this.container.FullName() + "/" + name [](start = 46, length = 38)

use string.Concat()

Comment (mikepizzo) [inline src/Microsoft.OData.Edm/Schema/EdmSingleton.cs:32]: So we are changing the "Path" property on a singleton to be container-qualified.  Where is this path used?

In general, we should NOT be using container-qualified paths.

File: test/FunctionalTests/Microsoft.OData.Edm.Tests/Csdl/CsdlWriterTests.cs [hunk 6]
```diff
@@ -1986,7 +1986,7 @@ public void CanWriteNavigationPropertyBindingWithTargetPathOnContainmentOnSingle
                     "<EntityContainer Name=\"Default\">" +
                        "<Singleton Name=\"Me\" Type=\"NS.Customer\" />" +
                        "<EntitySet Name=\"Customers\" EntityType=\"NS.Customer\">" +
-                         "<NavigationPropertyBinding Path=\"ContainedOrders/OrderLines\" Target=\"Me/ContainedOrderLines\" />" +
+                         "<NavigationPropertyBinding Path=\"ContainedOrders/OrderLines\" Target=\"NS.Default/Me/ContainedOrderLines\" />" +
```
Comment (mikepizzo) [inline test/FunctionalTests/Microsoft.OData.Edm.Tests/Csdl/CsdlWriterTests.cs:1989]: We should not be container-qualifying the target unless necessary (which it shouldn't be)

**Comment by mikepizzo** [comment]
Looking more closely, there really isn't ever a reason we should have to write the container-qualified form.  The container-qualified form is a leftover from when we had multiple entity containers.

When writing the target we should ideally always write the unqualified entityset/singleton name. If we're making a breaking change, I would move to writing without the container-qualifier for both entityset and singleton. If we're worried about breaking someone, we could add a flag to add the entitycontainer using the old format (no reason to add it in the new format).

I like the reading logic, which supports both the old (invalid) container-qualified format, the appropriate container-qualified format, and the unqualified format.

---
In reply to: [744196418](https://github.com/OData/odata.net/pull/1950#issuecomment-744196418) [](ancestors = 744196418)

**Review by mikepizzo — CHANGES_REQUESTED** [review]
:clock1:

**Comment by mikepizzo** [comment]
Looking more closely, there really isn't ever a reason we should have to write the container-qualified form.  The container-qualified form is a leftover from when we had multiple entity containers.

When writing the target we should ideally always write the unqualified entityset/singleton name. If we're making a breaking change, I would move to writing without the container-qualifier for both entityset and singleton. If we're worried about breaking someone, we could add a flag to add the entitycontainer using the old format (no reason to add it in the new format).

I like the reading logic, which supports both the old (invalid) container-qualified format, the appropriate container-qualified format, and the unqualified format.

**Comment by xuzhg** [comment]
I agreed to write the unqualifed entityset/singleton name.

---
In reply to: [745006913](https://github.com/OData/odata.net/pull/1950#issuecomment-745006913) [](ancestors = 745006913)

File: test/FunctionalTests/Microsoft.OData.Edm.Tests/Csdl/CsdlWriterTests.cs (hunk 6 above)
Comment (xuzhg) [reply test/FunctionalTests/Microsoft.OData.Edm.Tests/Csdl/CsdlWriterTests.cs:1989]: Revert to path without qualified name

---
In reply to: [542983048](https://github.com/OData/odata.net/pull/1950#discussion_r542983048) [](ancestors = 542983048)

File: src/Microsoft.OData.Edm/Schema/EdmSingleton.cs (hunk 5 above)
Comment (xuzhg) [reply src/Microsoft.OData.Edm/Schema/EdmSingleton.cs:32]: revert

---
In reply to: [542982488](https://github.com/OData/odata.net/pull/1950#discussion_r542982488) [](ancestors = 542982488)

Comment (xuzhg) [reply src/Microsoft.OData.Edm/Schema/EdmSingleton.cs:32]: Not applicable

---
In reply to: [542982068](https://github.com/OData/odata.net/pull/1950#discussion_r542982068) [](ancestors = 542982068)

File: src/Microsoft.OData.Edm/Schema/EdmEntitySet.cs (hunk 4 above)
Comment (xuzhg) [reply src/Microsoft.OData.Edm/Schema/EdmEntitySet.cs:44]: change to use the name directly, so not applicable

---
In reply to: [542981676](https://github.com/OData/odata.net/pull/1950#discussion_r542981676) [](ancestors = 542981676)

**Comment by xuzhg** [comment]
Yes, I agree to write the unqualified entityset/single name since we only have one entity container now.
I don't think we should introduce the flag because the old pattern is wrong pattern, and it only happens in containment binding.

---
In reply to: [744196418](https://github.com/OData/odata.net/pull/1950#issuecomment-744196418) [](ancestors = 744196418)

**Review by mikepizzo — APPROVED** [review]
:shipit:

**Pull request merged**